### DISCUSSION TITLE
Stripped down Status and StatusOr as a standalone library?

### DISCUSSION THREAD

**Discussion by automaciej** [abseil/abseil-cpp#1679] [Ideas]
Context: I'm writing some C++ code. My codebase is really small (7k SLOC), and I try to keep my dependencies lean; I build on Linux and macOS, potentially on Windows in the future. For now I only have SDL as a dependency. I don't use CMake (at the time being). I'd love to use Bazel, but I can't seem to build SDL using Bazel, so that's not an option for my project. For now I'm completely happy with a handwritten Makefile.

The main thing I'm missing is Status/StatusOr. The alternatives are kind of clunky; sometimes I use a bool as a return value whether the call was successful. Sometimes I pass in references, I'm experimenting with various approaches, but there really isn't a good substitute for Status and StatusOr.

I wrote a set of GNU Make rules to build Abseil, and I was able to include it in my app. But I realized that pulling in the entirety of Abseil as a dependency of my tiny program is a bit much; if Abseil doesn't build on macOS for example, for whatever reason, I will have a problem. 

I'd really like to use Status and StatusOr. I contemplated writing something similar myself, but if the work has already been done in Abseil, it makes a lot of sense to reuse it.

The header files currently pull in a number of dependencies that run relatively deep into the core of Abseil. I tried to identify the minimal set of dependencies, but didn't succeed. I ended up having to compile the entire Abseil.

Could Status and StatusOr be made more modular? Did anyone look into this?

**Comment by derekmauro** [comment]
We currently don't have any reason to look into making the `status` library more modular. Nearly all of our programs that use the `status` library also use its dependencies. We also don't want to guarantee what the dependencies of `status` are forever as it would hinder library evolution.

If this is really a problem for you, here are a couple ideas:
* Use `std::optional` instead.
* Use `std::expected` instead.

If you are having trouble building SDL under Bazel, here are some ideas:
* Use https://github.com/bazelbuild/rules_foreign_cc to invoke another build system from Bazel to build SDL.

**Comment by automaciej** [comment]
Hey, just wanted to thank you for your answer. While it's essentially a "no" to what I asked about, the suggestions for alternatives helped me figure out what to do.

I used `std::expected` as the return value, and for the error class, I just used a class enum. After some testing I realized I wanted to include a string message with my errors, and I asked ChatGPT to write me a lightweight substitute to `absl::Status`. That worked well enough.

If I were still at Google I'd have given you a peer bonus. Thanks again for such a kind and helpful answer.

**Comment by eugeneo** [comment]
Just in case someone needs it, `std::expected` is C++23 but `std::variant` is C++17. `std::variant<ReturnType, ErrorType>` is fairly close to `std::expected`.
